engines/mbic3.cpp

```diff
@@ -207,67 +208,25 @@ ProverResult ModelBasedIC3::step_0()
 
 bool ModelBasedIC3::intersects_bad()
 {
-  solver_->push();
-
-  // assert the last frame (conjunction over clauses)
-  assert_frame(reached_k_ + 1);
-
-  // see if it intersects with bad
-  solver_->assert_formula(bad_);
-
-  Result r = solver_->check_sat();
-  if (r.is_sat()) {
-    // create a proof goal for the bad state
-    const UnorderedTermSet & statevars = ts_.statevars();
-    TermVec cube_vec;
-    cube_vec.reserve(statevars.size());
-    Term eq;
-    for (auto sv : statevars) {
-      eq = solver_->make_term(Equal, sv, solver_->get_value(sv));
-      cube_vec.push_back(eq);
-    }
-
-    // reduce the bad cube
-    solver_->pop();
-    solver_->push();
-
-    Term neg_assert =
-      solver_->make_term(Not, solver_->make_term(And, bad_,
-                                                 get_frame(reached_k_ + 1)));
-    solver_->assert_formula(neg_assert);
-
-    TermVec splits;
-    split_eq(solver_, cube_vec, splits);
-
-    Term b;
-    TermVec bool_assump;
-    for (auto a : splits) {
-      unsigned i = 0;
-      b = label(a);
-      bool_assump.push_back(b);
-      solver_->assert_formula(solver_->make_term(Implies, b, a));
-    }
-
-    Result rr = solver_->check_sat_assuming(bool_assump);
-    assert(rr.is_unsat());
-
-    TermVec red_lits;
-    UnorderedTermSet core_set;
-    solver_->get_unsat_core(core_set);
-    for (size_t j = 0; j < bool_assump.size(); ++j) {
-      if (core_set.find(bool_assump[j]) != core_set.end()) {
-        red_lits.push_back(splits[j]);
-      }
-    }
-
-    Conjunction c(solver_, red_lits);
-    add_proof_goal(c, reached_k_ + 1);
+  TermVec conjuncts;
+  conjunctive_partition(bad_, conjuncts);
+
+  // include the last frame
+  // this is an optimization, it is not necessary for correctness
+  assert(frames_.size() == reached_k_ + 2);
+  for (auto c : frames_[reached_k_ + 1]) {
```

Comment:
perhaps we could use `frames_.back()`

Reply:
i don't have a strong preference for what I suggested

Reply:
Ahh good point! Made that change with 9e361e12d41017dda405d5735c49f1d0ff3f354a. I don't have a strong preference either. It seems kind of nice in principle to use it. Maybe we just keep it and if we have time we can evaluate whether it helps or not later?